docs/manpages/make_smbcodepage.1
.TH MAKE_SMBCODEPAGE 1 "23 Nov 2000" "make_smbcodepage 2.2.0-alpha1"
.PP 
.SH "NAME" 
make_codepage \- Construct a codepage file for Samba
.PP 
.SH "SYNOPSIS" 
.PP 
\fBmake_smbcodepage\fP [c|d] codepage inputfile outputfile
.PP 
.SH "DESCRIPTION" 
.PP 
This program is part of the \fBSamba\fP suite\&.
.PP 
\fBmake_smbcodepage\fP compiles or de-compiles codepage files for use
with the internationalization features of Samba 2\&.0
.PP 
.SH "OPTIONS" 
.PP 
.IP 
.IP "c|d" 
This tells \fBmake_smbcodepage\fP if it is compiling (\fBc\fP) a text
format code page file to binary, or (\fBd\fP) de-compiling a binary codepage
file to text\&.
.IP 
.IP "codepage" 
This is the codepage we are processing (a number, e\&.g\&. 850)\&.
.IP 
.IP "inputfile" 
This is the input file to process\&. In the \'\fBc\fP\' case this
will be a text codepage definition file such as the ones found in the
Samba \fIsource/codepages\fP directory\&. In the \'\fBd\fP\' case this will be the
binary format codepage definition file normally found in the
\fIlib/codepages\fP directory in the Samba install directory path\&.
.IP 
.IP "outputfile" 
This is the output file to produce\&.
.IP 
.PP 
.SH "Samba Codepage Files" 
.PP 
A text Samba codepage definition file is a description that tells
Samba how to map from upper to lower case for characters greater than
ascii 127 in the specified DOS code page\&.  Note that for certain DOS
codepages (437 for example) mapping from lower to upper case may be
non-symmetrical\&. For example, in code page 437 lower case a acute maps to
a plain upper case A when going from lower to upper case, but
plain upper case A maps to plain lower case a when lower casing a
character\&.
.PP 
A binary Samba codepage definition file is a binary representation of
the same information, including a value that specifies what codepage
this file is describing\&.
.PP 
As Samba does not yet use UNICODE (current for Samba version 2\&.0) you
must specify the client code page that your DOS and Windows clients
are using if you wish to have case insensitivity done correctly for
your particular language\&. The default codepage Samba uses is 850
(Western European)\&. Text codepage definition sample files are
provided in the Samba distribution for codepages 437 (USA), 737
(Greek), 850 (Western European) 852 (MS-DOS Latin 2), 861 (Icelandic),
866 (Cyrillic), 932 (Kanji SJIS), 936 (Simplified Chinese), 949
(Hangul) and 950 (Traditional Chinese)\&. Users are encouraged to write
text codepage definition files for their own code pages and donate
them to samba@samba\&.org\&. All codepage files in the
Samba \fIsource/codepages\fP directory are compiled and installed when a
\fI\'make install\'\fP command is issued there\&.
.PP 
The client codepage used by the \fBsmbd\fP server is
configured using the \fBclient code
page\fP parameter in the
\fBsmb\&.conf\fP file\&.
.PP 
.SH "FILES" 
.PP 
\fBcodepage_def\&.<codepage>\fP
.PP 
These are the input (text) codepage files provided in the Samba
\fIsource/codepages\fP directory\&.
.PP 
A text codepage definition file consists of multiple lines
containing four fields\&. These fields are : 
.PP 
.IP 
.IP o 
\fBlower\fP: which is the (hex) lower case character mapped on this
line\&.
.IP 
.IP o 
\fBupper\fP: which is the (hex) upper case character that the lower
case character will map to\&.
.IP 
.IP o 
\fBmap upper to lower\fP which is a boolean value (put either True
or False here) which tells Samba if it is to map the given upper case
character to the given lower case character when lower casing a
filename\&.
.IP 
.IP o 
\fBmap lower to upper\fP which is a boolean value (put either True
or False here) which tells Samba if it is to map the given lower case
character to the given upper case character when upper casing a
filename\&.
.IP 
.PP 
\fBcodepage\&.<codepage>\fP These are the output (binary) codepage files
produced and placed in the Samba destination \fIlib/codepage\fP
directory\&.
.PP 
.SH "INSTALLATION" 
.PP 
The location of the server and its support files is a matter for
individual system administrators\&. The following are thus suggestions
only\&.
.PP 
It is recommended that the \fBmake_smbcodepage\fP program be installed
under the \fI/usr/local/samba\fP hierarchy, in a directory readable by
all, writeable only by root\&. The program itself should be executable
by all\&.  The program should NOT be setuid or setgid!
.PP 
.SH "VERSION" 
.PP 
This man page is correct for version 2\&.0 of the Samba suite\&.
.PP 
.SH "SEE ALSO" 
.PP 
\fBsmb\&.conf(5)\fP, \fBsmbd (8)\fP
.PP 
.SH "AUTHOR" 
.PP 
The original Samba software and related utilities were created by
Andrew Tridgell samba@samba\&.org\&. Samba is now developed
by the Samba Team as an Open Source project similar to the way the
Linux kernel is developed\&.
.PP 
The original Samba man pages were written by Karl Auer\&. The man page
sources were converted to YODL format (another excellent piece of Open
Source software, available at
\fBftp://ftp\&.icce\&.rug\&.nl/pub/unix/\fP)
and updated for the Samba2\&.0 release by Jeremy Allison\&.
samba@samba\&.org\&.
.PP 
See \fBsamba (7)\fP to find out how to get a full
list of contributors and details on how to submit bug reports,
comments etc\&.


NEWS
Changes in release 0.3d

 * add krb5-config

 * fix a bug in 3des gss-api mechanism, making it compatible with the
   specification and the MIT implementation

 * make telnetd only allow a specific list of environment variables to
   stop it from setting `sensitive' variables

 * try to use an existing libdes

 * lib/krb5, kdc: use correct usage type for ap-req messages.  This
   should improve compatability with MIT krb5 when using 3DES
   encryption types

 * kdc: fix memory allocation problem

 * update config.guess and config.sub

 * lib/roken: more stuff implemented

 * bug fixes and portability enhancements

Changes in release 0.3c

 * lib/krb5: memory caches now support the resolve operation

 * appl/login: set PATH to some sane default

 * kadmind: handle several realms

 * bug fixes (including memory leaks)

Changes in release 0.3b

 * kdc: prefer default-salted keys on v5 requests

 * kdc: lowercase hostnames in v4 mode

 * hprop: handle more types of MIT salts

 * lib/krb5: fix memory leak

 * bug fixes

Changes in release 0.3a:

 * implement arcfour-hmac-md5 to interoperate with W2K

 * modularise the handling of the master key, and allow for other
   encryption types. This makes it easier to import a database from
   some other source without having to re-encrypt all keys.

 * allow for better control over which encryption types are created

 * make kinit fallback to v4 if given a v4 KDC

 * make klist work better with v4 and v5, and add some more MIT
   compatibility options

 * make the kdc listen on the krb524 (4444) port for compatibility
   with MIT krb5 clients

 * implement more DCE/DFS support, enabled with --enable-dce, see
   lib/kdfs and appl/dceutils

 * make the sequence numbers work correctly

 * bug fixes

Changes in release 0.2t:

 * bug fixes

Changes in release 0.2s:

 * add OpenLDAP support in hdb

 * login will get v4 tickets when it receives forwarded tickets

 * xnlock supports both v5 and v4

 * repair source routing for telnet

 * fix building problems with krb4 (krb_mk_req)

 * bug fixes

Changes in release 0.2r:

 * fix realloc memory corruption bug in kdc

 * `add --key' and `cpw --key' in kadmin

 * klist supports listing v4 tickets

 * update config.guess and config.sub

 * make v4 -> v5 principal name conversion more robust

 * support for anonymous tickets

 * new man-pages

 * telnetd: do not negotiate KERBEROS5 authentication if there's no keytab.

 * use and set expiration and not password expiration when dumping
   to/from ka server databases / krb4 databases

 * make the code happier with 64-bit time_t

 * follow RFC2782 and by default do not look for non-underscore SRV names

Changes in release 0.2q:

 * bug fix in tcp-handling in kdc

 * bug fix in expand_hostname

Changes in release 0.2p:

 * bug fix in `kadmin load/merge'

 * bug fix in krb5_parse_address

Changes in release 0.2o:

 * gss_{import,export}_sec_context added to libgssapi

 * new option --addresses to kdc (for listening on an explicit set of
   addresses)

 * bug fixes in the krb4 and kaserver emulation part of the kdc

 * other bug fixes

Changes in release 0.2n:

 * more robust parsing of dump files in kadmin
 * changed default timestamp format for log messages to extended ISO
   8601 format (Y-M-DTH:M:S)
 * changed md4/md5/sha1 APIes to be de-facto `standard'
 * always make hostname into lower-case before creating principal
 * small bits of more MIT-compatability
 * bug fixes

Changes in release 0.2m:

 * handle glibc's getaddrinfo() that returns several ai_canonname

 * new endian test

 * man pages fixes

Changes in release 0.2l:

 * bug fixes

Changes in release 0.2k:

 * better IPv6 test

 * make struct sockaddr_storage in roken work better on alphas

 * some missing [hn]to[hn]s fixed.

 * allow users to change their own passwords with kadmin (with initial
   tickets)

 * fix stupid bug in parsing KDC specification

 * add `ktutil change' and `ktutil purge'

Changes in release 0.2j:

 * builds on Irix

 * ftpd works in passive mode

 * should build on cygwin

 * work around broken IPv6-code on OpenBSD 2.6, also add configure
   option --disable-ipv6

Changes in release 0.2i:

 * use getaddrinfo in the missing places.

 * fix SRV lookup for admin server

 * use get{addr,name}info everywhere.  and implement it in terms of
   getipnodeby{name,addr} (which uses gethostbyname{,2} and
   gethostbyaddr)

Changes in release 0.2h:

 * fix typo in kx (now compiles)

Changes in release 0.2g:

 * lots of bug fixes:
   * push works
   * repair appl/test programs
   * sockaddr_storage works on solaris (alignment issues)
   * works better with non-roken getaddrinfo
   * rsh works
   * some non standard C constructs removed

Changes in release 0.2f:

 * support SRV records for kpasswd
 * look for both _kerberos and krb5-realm when doing host -> realm mapping

Changes in release 0.2e:

 * changed copyright notices to remove `advertising'-clause.
 * get{addr,name}info added to roken and used in the other code
   (this makes things work much better with hosts with both v4 and v6
    addresses, among other things)
 * do pre-auth for both password and key-based get_in_tkt
 * support for having several databases
 * new command `del_enctype' in kadmin
 * strptime (and new strftime) add to roken
 * more paranoia about finding libdb
 * bug fixes

Changes in release 0.2d:

 * new configuration option [libdefaults]default_etypes_des
 * internal ls in ftpd builds without KRB4
 * kx/rsh/push/pop_debug tries v5 and v4 consistenly
 * build bug fixes
 * other bug fixes

Changes in release 0.2c:

 * bug fixes (see ChangeLog's for details)

Changes in release 0.2b:

 * bug fixes
 * actually bump shared library versions

Changes in release 0.2a:

 * a new program verify_krb5_conf for checking your /etc/krb5.conf
 * add 3DES keys when changing password
 * support null keys in database
 * support multiple local realms
 * implement a keytab backend for AFS KeyFile's

source/utils/make_smbcodepage.c
/* 
   Unix SMB/Netbios implementation.
   Version 1.9.
   Create codepage files from codepage_def.XXX files.

   Copyright (C) Jeremy Allison 1997

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.
   
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.
   
   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

#include "includes.h"

static char *prog_name = NULL;

/*
 * Print program usage and die.
 */

void codepage_usage(const char *progname)
{
  fprintf(stderr, "Usage is : %s [c|d] <codepage> <inputfile> <outputfile>\n",
         progname);
  exit(1);
}

/*
 * Read a line from a buffer into a line buffer. Ensure null
 * terminated.
 */

void read_line( char **buf, char *line_buf, int size)
{
  char *p = *buf;
  int num = 0;

  for(; *p && (*p != '\n'); p++)
  {
    if(num < (size - 1))
      line_buf[num++] = *p;
  }
  if(*p)
    p++; /* Go past the '\n' */
  line_buf[num] = '\0';
  *buf = p;
}

/*
 * Strip comment lines and blank lines from the data.
 * Copies into a new buffer and frees the old.
 * Returns the number of lines copied.
 */

int clean_data( char **buf, uint32 *size)
{
  char linebuf[512];
  char *p = *buf;
  int num_lines = 0;
  char *newbuf = (char *)malloc( *size + 1);
  char *newbuf_p = NULL;

  if(newbuf == NULL)
  {
    fprintf(stderr, "%s: malloc fail for size %d.\n", prog_name, *size + 1);
    exit(1);
  }

  newbuf_p = newbuf;
  *newbuf_p = '\0';

  while( *p )
  {
    char *cp;

    read_line( &p, linebuf, sizeof(linebuf));
    /* Null terminate after comment. */
    if((cp = strchr( linebuf, '#'))!= NULL)
      *cp = '\0';

    for(cp = linebuf;*cp && isspace(*cp); cp++)
      ;

    if(*cp == '\0')
      continue;

    strcpy(newbuf_p, cp);
    num_lines++;
    newbuf_p += (strlen(newbuf_p) + 1);
  }

  free(*buf);
  *buf = newbuf;
  return num_lines;
}

/*
 * Parse a byte from a codepage file.
 */

BOOL parse_byte(char *buf, unsigned char *bp)
{
  unsigned int b;
  char *endptr = NULL;

  b = (unsigned int)strtol(buf, &endptr, 0);
  if(endptr == buf || b > 255)
    return False;

  *bp = (unsigned char)b;
  return True;
}

/*
 * Parse a bool from a codepage file.
 */

BOOL parse_bool(char *buf, unsigned char *bp)
{
  if(isdigit(*buf))
  {
    char *endptr = NULL;

    *bp = (unsigned char)strtol(buf, &endptr, 0);
    if(endptr == buf )
      return False;
    if(*bp != 0)
      *bp = 1;
  } else {
    if(strcasecmp(buf, "True") && strcasecmp(buf, "False"))
      return False;
    if(strcasecmp(buf, "True")==0)
      *bp = 1;
    else
      *bp = 0;
  }
  return True;
}

/*
 * Print a parse error and exit.
 */

void parse_error(char *buf, char *msg)
{
  fprintf(stderr, "%s: %s whilst parsing line \n%s\n", prog_name,
          msg, buf);
  exit(1);
}
    
/*
 * Create a compiled codepage file from a codepage definition file.
 */

int do_compile(int codepage, const char *input_file, const char *output_file)
{
  FILE *fp = NULL;
  uint32 size = 0;
  char *buf = NULL;
  char output_buf[CODEPAGE_HEADER_SIZE + 512];
  int num_lines = 0;
  int i = 0;
  struct stat st;

  /* Get the size of the input file. Read the entire thing into memory. */
  if(stat((char *)input_file, &st)!= 0)
  {
    fprintf(stderr, "%s: failed to get the file size for file %s. Error was %s\n",
            prog_name, input_file, strerror(errno));
    exit(1);
  }

  size = (uint32)st.st_size;

  /* I don't believe these things should be bigger than 100k :-) */
  if(size > 100*1024)
  {
    fprintf(stderr, "%s: filesize %d is too large for a codepage definition file. \
The maximum size I will believe is 100k.\n", prog_name, size);
    exit(1);
  }

  if((fp = fopen(input_file, "r")) == NULL)
  {
    fprintf(stderr, "%s: cannot open file %s for input.\n", prog_name, input_file);
    exit(1);
  }

  /* As we will be reading text, allocate one more byte for a '\0' */
  if((buf = (char *)malloc( size + 1 )) == NULL)
  {
    fprintf(stderr, "%s: malloc fail for size %d.\n", prog_name, size + 1);
    fclose(fp);
    exit(1);
  }

  if(fread( buf, 1, size, fp) != size)
  {
    fprintf(stderr, "%s: read failed for file %s. Error was %s.\n", prog_name,
            input_file, strerror(errno));
    free((char *)buf);
    fclose(fp);
    exit(1);
  }

  /* Null terminate the text read. */
  buf[size] = '\0';

  /* Go through the data line by line, strip out comments (anything
     after a '#' to end-of-line) and blank lines. The rest should be
     the codepage data.
   */

  num_lines = clean_data( &buf, &size);

  /* There can be a maximum of 128 lines. */
  if(num_lines > 128)
  {
    fprintf(stderr, "%s: There can be a maximum 128 lines of data in a codepage \
definition file. File %s has %d.\n", prog_name, input_file, num_lines);
    exit(1);
  }

  /* Setup the output file header. */
  SSVAL(output_buf,CODEPAGE_VERSION_OFFSET,CODEPAGE_FILE_VERSION_ID);
  SSVAL(output_buf,CODEPAGE_CLIENT_CODEPAGE_OFFSET,(uint16)codepage);
  SIVAL(output_buf,CODEPAGE_LENGTH_OFFSET,(num_lines * 4));

  /* Now convert the lines into the compiled form. */
  for(i = 0; i < num_lines; i++)
  {
    char token_buf[512];
    char *p = buf;
    unsigned char b = 0;

    /* Get the 'lower' value. */
    if(!next_token(&p, token_buf, NULL))
      parse_error(buf, "cannot parse first value");
    if(!parse_byte( token_buf, &b))
      parse_error(buf, "first value doesn't resolve to a byte");

    /* Add this to the output buffer. */
    SCVAL(output_buf,CODEPAGE_HEADER_SIZE+(i*4),b);

    /* Get the 'upper' value. */
    if(!next_token(&p, token_buf, NULL))
      parse_error(buf, "cannot parse second value");
    if(!parse_byte( token_buf, &b))
      parse_error(buf, "second value doesn't resolve to a byte");

    /* Add this to the output buffer. */
    SCVAL(output_buf,CODEPAGE_HEADER_SIZE+(i*4) + 1,b);

    /* Get the 'upper to lower' value. */
    if(!next_token(&p, token_buf, NULL))
      parse_error(buf, "cannot parse third value");
    if(!parse_bool( token_buf, &b))
      parse_error(buf, "third value doesn't resolve to a boolean");

    /* Add this to the output buffer. */
    SCVAL(output_buf,CODEPAGE_HEADER_SIZE+(i*4) + 2,b);

    /* Get the 'lower to upper' value. */
    if(!next_token(&p, token_buf, NULL))
      parse_error(buf, "cannot parse fourth value");
    if(!parse_bool( token_buf, &b))
      parse_error(buf, "fourth value doesn't resolve to a boolean");

    /* Add this to the output buffer. */
    SCVAL(output_buf,CODEPAGE_HEADER_SIZE+(i*4) + 3,b);

    buf += (strlen(buf) + 1);
  }

  /* Now write out the output_buf. */
  if((fp = fopen(output_file, "w"))==NULL)
  {
    fprintf(stderr, "%s: Cannot open output file %s. Error was %s.\n",
            prog_name, output_file, strerror(errno));
    exit(1);
  }

  if(fwrite(output_buf, 1, CODEPAGE_HEADER_SIZE + (num_lines*4), fp) !=
         CODEPAGE_HEADER_SIZE + (num_lines*4))
  {
    fprintf(stderr, "%s: Cannot write output file %s. Error was %s.\n",
            prog_name, output_file, strerror(errno));
    exit(1);
  }  

  fclose(fp);

  return 0;
}

/*
 * Placeholder for now.
 */

int do_decompile( int codepage, const char *input_file, const char *output_file)
{
  uint32 size = 0;
  struct stat st;
  char header_buf[CODEPAGE_HEADER_SIZE];
  char *buf = NULL;
  FILE *fp = NULL;
  int num_lines = 0;
  int i = 0;

  /* Get the size of the input file. Read the entire thing into memory. */
  if(stat((char *)input_file, &st)!= 0)
  {
    fprintf(stderr, "%s: failed to get the file size for file %s. Error was %s\n",
            prog_name, input_file, strerror(errno));
    exit(1);
  }

  size = (uint32)st.st_size;

  if( size < CODEPAGE_HEADER_SIZE || size > (CODEPAGE_HEADER_SIZE + 256))
  { 
    fprintf(stderr, "%s: file %s is an incorrect size for a \
code page file.\n", prog_name, input_file);
    exit(1);
  } 
  
  /* Read the first 8 bytes of the codepage file - check
     the version number and code page number. All the data
     is held in little endian format.
   */
    
  if((fp = fopen( input_file, "r")) == NULL)
  { 
    fprintf(stderr, "%s: cannot open file %s. Error was %s\n",
              prog_name, input_file, strerror(errno));
    exit(1);
  } 
  
  if(fread( header_buf, 1, CODEPAGE_HEADER_SIZE, fp)!=CODEPAGE_HEADER_SIZE)
  { 
    fprintf(stderr, "%s: cannot read header from file %s. Error was %s\n",
              prog_name, input_file, strerror(errno));
    exit(1);
  } 
  
  /* Check the version value */
  if(SVAL(header_buf,CODEPAGE_VERSION_OFFSET) != CODEPAGE_FILE_VERSION_ID)
  { 
    fprintf(stderr, "%s: filename %s has incorrect version id. \
Needed %hu, got %hu.\n",
          prog_name, input_file, (uint16)CODEPAGE_FILE_VERSION_ID, 
          SVAL(header_buf,CODEPAGE_VERSION_OFFSET));
    exit(1);
  } 
  
  /* Check the codepage matches */
  if(SVAL(header_buf,CODEPAGE_CLIENT_CODEPAGE_OFFSET) != (uint16)codepage)
  { 
    fprintf(stderr, "%s: filename %s has incorrect codepage. \
Needed %hu, got %hu.\n",
           prog_name, input_file, (uint16)codepage, 
           SVAL(header_buf,CODEPAGE_CLIENT_CODEPAGE_OFFSET));
    exit(1);
  } 
  
  /* Check the length is correct. */
  if(IVAL(header_buf,CODEPAGE_LENGTH_OFFSET) !=
                 (unsigned int)(size - CODEPAGE_HEADER_SIZE))
  { 
    fprintf(stderr, "%s: filename %s has incorrect size headers. \
Needed %u, got %u.\n", prog_name, input_file, size - CODEPAGE_HEADER_SIZE,
               IVAL(header_buf,CODEPAGE_LENGTH_OFFSET));
    exit(1);
  } 
  
  size -= CODEPAGE_HEADER_SIZE; /* Remove header */
    
  /* Make sure the size is a multiple of 4. */
  if((size % 4 ) != 0)
  { 
    fprintf(stderr, "%s: filename %s has a codepage size not a \
multiple of 4.\n", prog_name, input_file);
    exit(1);
  } 
  
  /* Allocate space for the code page file and read it all in. */
  if((buf = (char *)malloc( size )) == NULL)
  { 
    fprintf (stderr, "%s: malloc fail for size %d.\n",
             prog_name, size );
    exit(1);
  } 
  
  if(fread( buf, 1, size, fp)!=size)
  { 
    fprintf(stderr, "%s: read fail on file %s. Error was %s.\n",
              prog_name, input_file, strerror(errno));
    exit(1);
  } 
  
  fclose(fp);
    
  /* Now dump the codepage into an ascii file. */
  if((fp = fopen(output_file, "w")) == NULL)
  {
    fprintf(stderr, "%s: cannot open file %s. Error was %s\n",
              prog_name, output_file, strerror(errno));
    exit(1);
  } 

  fprintf(fp, "#\n# Codepage definition file for IBM Code Page %d.\n#\n",
          codepage);
  fprintf(fp, "# This file was automatically generated.\n#\n");
  fprintf(fp, "# defines lower->upper mapping.\n");
  fprintf(fp, "#\n#The columns are :\n# lower\tupper\tu-t-l\tl-t-u\n#\n");

  num_lines = size / 4;
  for( i = 0; i < num_lines; i++)
  {
    fprintf(fp, "0x%02X\t0x%02X\t%s\t%s\n", CVAL(buf, (i*4)), CVAL(buf, (i*4)+1),
           CVAL(buf, (i*4)+2) ? "True" : "False",
           CVAL(buf, (i*4)+3) ? "True" : "False");
  }
  fclose(fp);
  return 0;
}

int main(int argc, char **argv)
{
  int codepage = 0;
  char *input_file = NULL;
  char *output_file = NULL;
  BOOL compile = False;

  prog_name = argv[0];

  if(argc != 5)
    codepage_usage(prog_name);

  if(argv[1][0] != 'c' && argv[1][0] != 'C' && argv[1][0] != 'd' &&
     argv[1][0] != 'D')
    codepage_usage(prog_name);

  input_file = argv[3];
  output_file = argv[4];

  /* Are we compiling or decompiling. */
  if(argv[1][0] == 'c' || argv[1][0] == 'C')
    compile = True;
 
  /* Convert the second argument into a client codepage value. */
  if((codepage = atoi(argv[2])) == 0)
  {
    fprintf(stderr, "%s: %s is not a valid codepage.\n", prog_name, argv[2]);
    exit(1);
  }

  if(compile)
    return do_compile( codepage, input_file, output_file);
  else
    return do_decompile( codepage, input_file, output_file);
}
